Add NegativeLogLikelihood(NllLoss) op (#2551)
onnx/onnx: docs/Changelog.md

<dd>Constrain index tensor to int64</dd>
</dl>

> ### <a name="NllLoss-12"></a>**NllLoss-12**</a>

Reviewer: The generated docs don't say this is operator has a function, because [gen_doc.py](https://github.com/onnx/onnx/blob/master/onnx/defs/gen_doc.py#L204-L207) only checks for `has_function`. It also isn't listed in the table of contents as one of the operators with a function registered.

>   A NllLoss op computes (weighted) negative log likelihood loss.

Reviewer 2: ```suggestion
  A NegativeLogLikelihood op computes (weighted) negative log likelihood.
```

>   Its "input" tensor has the shape of (N, C, d1, d2, ..., dk) where k >= 0. The "input" tensor contains log-probabilities for samples being in a class of [0, C).

Reviewer 2: ```suggestion
  Its "input" tensor has the shape of (N, C, d0, d1, ..., d_{k-1}) where k >= 0. The "input" tensor contains log-probabilities for samples being in a class of [0, C).
```
but the `k` here starts with 0.

>   Its "input" tensor has the shape of (N, C, d1, d2, ..., dk) where k >= 0. The "input" tensor contains log-probabilities for samples being in a class of [0, C).

Reviewer 2: What is the log-probabilities for a sample? Is it `input[i, :, k, l, m, ...]`?

>   The op's "target" input tensor has the shape of (N, d1, d2, ..., dk). It contains classifications (one of C classes) for N x d1 x d2 x ... x dk samples. 

Reviewer 2: ```suggestion
  The operator's "target" input tensor has the shape of (N, d1, d2, ..., dk). It contains classifications (one of C classes) for N x d1 x d2 x ... x dk samples. 
```
Let's use operator when we can.

>   The op's "target" input tensor has the shape of (N, d1, d2, ..., dk). It contains classifications (one of C classes) for N x d1 x d2 x ... x dk samples. 

Reviewer 2: cla
```suggestion
  The op's "target" input tensor has the shape of (N, d1, d2, ..., dk). It encodes class labels (one of C classes) for N x d1 x d2 x ... x dk samples. 
```

>   The loss for a sample at n, d_1, d_2,...d_k being classified as class c = target[n][d_1][d_2]...[d_k] is computed as:

Reviewer 2: ```suggestion
  The loss for a sample at n, d_1, d_2,...d_k being classified as class c = target[n][d_1][d_2]...[d_k] is computed as:
```
`a sample at n, d_1, d_2,...d_k` is not a formal description. Do you mean `input[n, d_1, d_2, ..., d_k]`?

  
      loss[n][d_1][d_2]...[d_k] = -input[n][c][d_1][d_2]...[d_k].
  
  When an optional "weight" is provided, the sample loss is calculated as:
  
      loss[n][d_1][d_2]...[d_k] = -input[n][c][d_1][d_2]...[d_k] * weight[c].
  

>   If "reduction" attribute is set to "none", the op's output will be the above loss with shape (N, d1, d2, ..., dk).

Reviewer 2: ```suggestion
  If "reduction" attribute is set to "none", the operator's output will be the above loss with shape (N, d1, d2, ..., dk).
```

              [[0.0, 1.0], [2.0, 2.0], [1.0, 2]]]
      target = [[2, 1], [0, 2]]
      weight = [0.2, 0.3, 0.1]
      loss = np.zeros((N, d1))
      for n in range(N):
          for d_1 in range(d1):
              c = target[n][d_1]
              loss[n][d_1] = -input[n][c][d_1] * weight[c]
  
      loss = np.sum(loss)
      # print(loss)
      # -1.1
  
  Example 3:
  
      # weighted negative log likelihood loss, mean reduction
      N, C, d1 = 2, 3, 2
      input = [[[1.0, 2.0], [2.0, 2.0], [3.0, 2.0]], 
              [[0.0, 1.0], [2.0, 2.0], [1.0, 2]]]
      target = [[2, 1], [0, 2]]
      weight = [0.2, 0.3, 0.1]
      loss = np.zeros((N, d1))
      weight_total = 0
      for n in range(N):
          for d_1 in range(d1):
              c = target[n][d_1]
              loss[n][d_1] = -input[n][c][d_1] * weight[c]
              weight_total = weight_total + weight[c]
  
      loss = np.sum(loss) / weight_total
      # print(loss)
      # -1.57

#### Version

This version of the operator has been available since version 12 of the default ONNX operator set.

#### Attributes

<dl>
<dt><tt>reduction</tt> : string (default is mean)</dt>

> <dd>Type of reduction to apply to loss: none, sum, mean(default). 'none': the output is the loss for each sample in the batch.'sum': the output will be summed. 'mean': the sum of the output will be divided by the batch_size.</dd>

Reviewer 2: ```suggestion
<dd>Type of reduction to apply to loss: none, sum, mean (default). 'none': the output is the loss for each sample in the batch. 'sum': the output will be summed into a scalar. 'mean': similar to the sum mode, but further divided by the batch_size.</dd>
```
Also, what is `batch_size`? I can't find its definition in the doc.

Author: added shape tests

<dd>Constrain index tensor to int64</dd>
</dl>

### <a name="NegativeLogLikelihoodLoss-12"></a>**NegativeLogLikelihoodLoss-12**</a>

  A NegativeLogLikelihoodLoss operator computes (weighted) negative log likelihood loss.
  Its "input" tensor has the shape of (N, C, d1, d2, ..., dk) where k >= 0.

>   The "input" tensor contains log-probabilities for input[n, :, d_1, d_2,...d_k] being in a class of [0, C).

Reviewer 2: ```suggestion
  The "input" tensor contains log-probabilities for input[n, :, d_1, d_2,..., d_k] being in a class of [0, C).
```

>   The operator's "target" input tensor has the shape of (N, d1, d2, ..., dk). It encode class labels (one of C classes) for N x d1 x d2 x ... x dk samples.

Reviewer 2: ```suggestion
  The operator's "target" input tensor has the shape of (N, d1, d2, ..., dk). It encodes class labels (one of C classes) for N x d1 x d2 x ... x dk samples.
```

>   The loss value for input[n, :, d_1, d_2,...d_k] being classified as class c = target[n][d_1][d_2]...[d_k] is computed as:

Reviewer 2: ```suggestion
  The loss value for input[n, :, d_1, d_2,..., d_k] being classified as class c = target[n][d_1][d_2]...[d_k] is computed as:
```
Copy-and-paste? :)

Author: fixed

              [[0.0, 1.0], [2.0, 2.0], [1.0, 2]]]
      target = [[2, 1], [0, 2]]
      weight = [0.2, 0.3, 0.1]
      loss = np.zeros((N, d1))
      for n in range(N):
          for d_1 in range(d1):
              c = target[n][d_1]
              loss[n][d_1] = -input[n][c][d_1] * weight[c]
  
      loss = np.sum(loss)
      # print(loss)
      # -1.1
  
  Example 3:
  
      # weighted negative log likelihood loss, mean reduction
      N, C, d1 = 2, 3, 2
      input = [[[1.0, 2.0], [2.0, 2.0], [3.0, 2.0]], 
              [[0.0, 1.0], [2.0, 2.0], [1.0, 2]]]
      target = [[2, 1], [0, 2]]
      weight = [0.2, 0.3, 0.1]
      loss = np.zeros((N, d1))
      weight_total = 0
      for n in range(N):
          for d_1 in range(d1):
              c = target[n][d_1]
              loss[n][d_1] = -input[n][c][d_1] * weight[c]
              weight_total = weight_total + weight[c]
  
      loss = np.sum(loss) / weight_total
      # print(loss)
      # -1.57

#### Version

This version of the operator has been available since version 12 of the default ONNX operator set.

#### Attributes

<dl>
<dt><tt>reduction</tt> : string (default is mean)</dt>

> <dd>Type of reduction to apply to loss: none, sum, mean(default). 'none': the output is the loss for each sample.'sum': the output will be summed. 'mean': the sum of the output will be divided by the sum of applied weights.</dd>

Reviewer 2: ```suggestion
<dd>Type of reduction to apply to loss: none, sum, mean (default). 'none': the output is the loss for each sample.'sum': the output will be summed. 'mean': the sum of the output will be divided by the sum of applied weights.</dd>
```
